src/utils/packageUtils.ts

```diff
@@ -0,0 +1,976 @@
+  const ids = [...new Set<string>(subscriberPackageVersionIds)];
+
+  const query =
+    'SELECT SubscriberPackageVersionId, MajorVersion, MinorVersion, PatchVersion, BuildNumber FROM Package2Version WHERE SubscriberPackageVersionId IN (%IDS%)';
+
+  const records = await queryWithInConditionChunking<PackageVersionString>(query, ids, '%IDS%', connection);
+  if (records && records.length > 0) {
+    results = new Map<string, string>(
+      records.map((record) => {
+        const version = concatVersion(
+          record.MajorVersion,
+          record.MinorVersion,
+          record.PatchVersion,
+          record.BuildNumber
+        );
+        return [record.SubscriberPackageVersionId, version];
+      })
+    );
+  }
+  return results;
+}
+/**
+ * For queries with an IN condition, determine if the WHERE clause will exceed
+ * SOQL's 4000 character limit.  Perform multiple queries as needed to stay below the limit.
+ *
+ * @return concatenated array of records returned from the resulting query(ies)
+ * @param query The full query to execute containing the replaceToken param in its IN clause
+ * @param items The IN clause items.  A length-appropriate single-quoted comma-separated string chunk will be made from the items.
+ * @param replaceToken A placeholder in the query's IN condition that will be replaced with the chunked items
+ * @param connection For tooling query
+ */
+export async function queryWithInConditionChunking<T = Record<string, unknown>>(
+  query: string,
+  items: string[],
+  replaceToken: string,
+  connection: Connection
+) {
+  let records: T[] = [];
+  if (!query || !items || !replaceToken) {
+    return records;
+  }
+
+  const whereClause = query.substring(query.toLowerCase().indexOf('where'), query.length);
+  const inClauseItemsMaxLength = SOQL_WHERE_CLAUSE_MAX_LENGTH - whereClause.length - replaceToken.length;
+
+  let itemsQueried = 0;
+  while (itemsQueried < items.length) {
+    const chunkCount: number = getInClauseItemsCount(items, itemsQueried, inClauseItemsMaxLength);
+    if (chunkCount === 0) {
+      throw messages.createError('itemDoesNotFitWithinMaxLength', [
+        query,
+        items[itemsQueried].slice(0, 20),
+        items[itemsQueried].length,
+        inClauseItemsMaxLength,
+      ]);
+    }
+    const itemsStr = `${items.slice(itemsQueried, itemsQueried + chunkCount).join("','")}`;
+    const queryChunk = query.replace(replaceToken, itemsStr);
+    const result = await connection.tooling.query<T>(queryChunk);
+    if (result && result.records.length > 0) {
+      records = records.concat(result.records);
+    }
+    itemsQueried += chunkCount;
+  }
+  return records;
+}
+/**
+ *   Returns the number of items that can be included in a quoted comma-separated string (e.g., "'item1','item2'") not exceeding maxLength
+ */
+// TODO: this function cannot handle a single item that is longer than maxLength - what to do, since this could be the root cause of an infinite loop?
```

Comment:
Not sure if this function should throw when an item count of zero will be returned. I modified the calling function to throw, just so we don't get into infinite loop (very unlikely, it would take first item to have a length around 4000 to trigger the loop.